### PR TITLE
Initial info tree and material=>recycling info conversion object

### PR DESCRIPTION
**Pull request by adamosimpson21**
nodeList and whereDoesItGo files only contain Array of objects used in infoTreeLogic page. That page converts one of these objects (currently the one housed in whereDoesItGo) and converts it into a tree structure, with child=>parent and parent=> child relationships. I've left comments in the code describing the specific functions and progress that can be made in the future

### CODE REVIEW

File: infoTree/nodeList.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+//I believe this is the list of every branch with category titles, but it's too late for me to really look at it right now.
```
Comment (AndyCErnst) [inline infoTree/nodeList.js]: The "it's too late for me to really look at it right now." seems like it should be a PR comment, not a code comment

Comment (AndyCErnst) [reply infoTree/nodeList.js]: Also, you don't mention that this list is from seattle.gov. Right now I wouldn't know if this is the raw or cleaned data.

File: infoTree/recycleBuddyScratch.js [hunk 2]
```diff
@@ -0,0 +1,200 @@
+//import the current list of leaves from separate file
```
Comment (AndyCErnst) [inline infoTree/recycleBuddyScratch.js]: comments describing how javascript works aren't useful. I can see where you might want more of this kind of comment in a project like this, but `import` is easy enough to look up if people don't understand, and it's already used in most files.

File: infoTree/recycleBuddyScratch.js [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+import whereDoesItGo from './whereDoesItGo';
+import nodeList from './nodeList'
+
+//I found this on a github repo (under one of my starred Repos) so obviously it's not my own work
```
Comment (AndyCErnst) [inline infoTree/recycleBuddyScratch.js]: link to the repo, give credit where it's due. I'd shorten the comment to something like:
`Tree implementation originally from <repo>. Trimmed down from the original version.`

File: infoTree/recycleBuddyScratch.js [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+//TODO: make this work
+Tree.prototype.findAddress = function(data){
+	if(!this.contains(data, this.traverseBF)){
+		return console.log("Tree does not contain "+ data);
+	}
+	return true
+}
+
+
+//formatting whereRecycleProto. Changes the value string to an array.
```
Comment (AndyCErnst) [inline infoTree/recycleBuddyScratch.js]: This looks like where your work starts. I'd like to see a clean separation between code that we don't own and code that we do. If we are heavily modifying this we might consider it _ours_ at some point, but for now I'd much rather see it as a library installed with npm, or in a vendor folder. If you are modifying it, I want to see your modifications all in the same file so it's obvious what's going on.

File: infoTree/recycleBuddyScratch.js [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+
+//formatting whereRecycleProto. Changes the value string to an array. 
+//This could be done with regular expressions, but an array containing the parents seems easier to me.
+//TODO: adapt for different inputs. because of the slice(4,-1) this could break easily when we include the category titles
+function formatRecycleObj(whereRecycleObj){
+  for(let i = 0; i<whereRecycleObj.length; i++){
+  	//splits into array and slices from WhereDoesItGo/[......]/index.htm
+    whereRecycleObj[i].value = whereRecycleObj[i].value.split("/").slice(4, -1)
```
Comment (AndyCErnst) [inline infoTree/recycleBuddyScratch.js]: clean code: avoid traditional for loops. This would be clearer in a forEach.

File: infoTree/recycleBuddyScratch.js [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+//each of those branches can have more branches with more specific recycling materials/objects
+
+//initialize tree with first node names rootNode
+var recycleTree = new Tree()
+recycleTree.add("rootNode")
+
+function createRecycleTree(recycleArray){
+  //format object so its easier to use if it's not formatted
+  if(typeof recycleArray[0].value==="string"){
+    recycleArray = formatRecycleObj(recycleArray);
```
Comment (AndyCErnst) [inline infoTree/recycleBuddyScratch.js]: nice that this data cleaner is its own function.

File: infoTree/recycleBuddyScratch.js [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+  }
+  //loop through every object in recycleArray 
+  for(let i=0; i<recycleArray.length; i++){
+    //for every branch, this creates the branch structure, but no the leaves
+    for(let j=0; j<recycleArray[i].value.length; j++){
+      //check to see if the branch is in the tree already
+      if(!recycleTree.contains(recycleArray[i].value[j], recycleTree.traverseBF)){
+        // console.log(recycleArray[i].value[j])
+        // checks to make sure the branch's parent isn't the rootNode
+        if(recycleArray[i].value[j-1]){
```
Comment (AndyCErnst) [inline infoTree/recycleBuddyScratch.js]: forEach loops will clear up a lot of `i`s and `j`s. The value of `recycleArray[i].value[j-1]` is particularly hard to understand, even with your comment

File: infoTree/whereDoesItGo.js [hunk 8]
```diff
@@ -0,0 +1,826 @@
+//this list only contains the leaves of the tree structure
```
Comment (AndyCErnst) [inline infoTree/whereDoesItGo.js:1]: be more explicit, "These are the leaves of a tree generated from the Seattle.gov recycling website using recycleBuddyScratch.js"

Comment (AndyCErnst) [reply infoTree/whereDoesItGo.js:1]: Although in the end, I don't imagine we are keeping this file. This is just to feed the crawler right?

File: infoTree/recycleBuddyScratch.js [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+    }
+    //adds the leaves 
+    recycleTree.add(recycleArray[i].label, recycleArray[i].value[recycleArray[i].value.length-1]);
+  }
+  // console.log(recycleTree);
+  return recycleTree
+}
+
+//call the function
+createRecycleTree(whereDoesItGo);
```
Comment (AndyCErnst) [inline infoTree/recycleBuddyScratch.js]: seems silly to call the function then throw away the result

File: infoTree/recycleBuddyScratch.js [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+  //format object so its easier to use if it's not formatted
+  if(typeof recycleArray[0].value==="string"){
+    recycleArray = formatRecycleObj(recycleArray);
+  }
+  //loop through every object in recycleArray 
+  for(let i=0; i<recycleArray.length; i++){
+    //for every branch, this creates the branch structure, but no the leaves
+    for(let j=0; j<recycleArray[i].value.length; j++){
+      //check to see if the branch is in the tree already
+      if(!recycleTree.contains(recycleArray[i].value[j], recycleTree.traverseBF)){
```
Comment (AndyCErnst) [inline infoTree/recycleBuddyScratch.js]: their `contains` doesn't take a second parameter.

**Review by AndyCErnst — CHANGES_REQUESTED** [review]
Eventually, we should delete the data files that won't go in the final version. We should at least put them in a data folder so it's clear they aren't code.

Nice job tackling this. This is probably the most data structure heavy part of the app we'll need to build

File: infoTree/recycleBuddyScratch.js (hunk 3 above)
Comment (adamosimpson21) [reply infoTree/recycleBuddyScratch.js]: Cool. I didn't really know how to handle this and discuss it through comments. That's much better than what I did. will change

File: infoTree/recycleBuddyScratch.js (hunk 9 above)
Comment (adamosimpson21) [reply infoTree/recycleBuddyScratch.js]: yea, I was mostly using it to view what I had in the console, as we're not using it anywhere yet. Eventually It'll be written to a file or imported in some way

File: infoTree/createDataStore.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+	.then(html => scrapeHtml(html))
+	.catch(err => {
+		console.log("Could not connect to site: " + err.message)
+	})
+}
+
+function scrapeHtml(html){
+	var howToDispose = {};
+	//Input: HTML from seattle.gov where does it go site
+	//Returns an object that holds formatted data as described in seattleGovScrape.js
```
Comment (AndyCErnst) [inline infoTree/createDataStore.js]: comments like this are more typically in javadoc style, something like
```
 /**
 * Simplifies and normalizes scraped html into a usable datastructure.
 * 
 * @param html   scraped html from Seattle.gov
 * @return tree of html nodes 
 */
```

File: infoTree/createRecycleTree.js [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+
+Tree.prototype.findEveryLeaf = function(){
+	var leafArray = []
+	//TODO: find every leaf Node and push it into leafArray
+	return leafArray;
+}
+
+//finds the URL to scrape data from by adding parent's name to base URL
+Tree.prototype.findUrl = function(leaf){
+	url = "https://www.seattle.gov/util/MyServices/WhereDoesItGo/"
```
Comment (AndyCErnst) [inline infoTree/createRecycleTree.js]: no variable declaration for `url`, this makes it a global which we definitely want to avoid.

File: infoTree/createRecycleTree.js [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+
+
+
+
+
+//formatting whereDoesItGo. Changes the value string => array. 
+//This could be done with regular expressions, but an array containing the parents seems easier to me.
+//TODO: adapt for different inputs. because of the slice(4,-1) this could break easily when we include the category titles
+function formatRecycleArray(whereRecycleArray){
+  whereRecycleArray.forEach(branchInfo => branchInfo.value = branchInfo.value.split("/").slice(4,-1))
```
Comment (AndyCErnst) [inline infoTree/createRecycleTree.js]: I'd break `branchInfo.value.split("/").slice(4,-1)` into its own function with a descriptive name

File: infoTree/createRecycleTree.js [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+//creates a tree with a root node of "rootNode" with children of general recycle materials based on "Where does it go" on seattle.gov
+//each of those branches can have more branches with more specific recycling materials/objects
+export default function createRecycleTree(recycleArray){
+  var recycleTree = new Tree()
+  recycleTree.add("rootNode")
+
+  //format object so its easier to use if it's not formatted
+  (typeof recycleArray[0].value==="string") && (recycleArray = formatRecycleArray(recycleArray))
```
Comment (AndyCErnst) [inline infoTree/createRecycleTree.js]: what is going on here? It seems like this wants to be 
```
if (typeof recycleArray[0].value==="string") {
   recycleArray = formatRecycleArray(recycleArray);
}
```

File: infoTree/createRecycleTree.js [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+
+  recycleArray.forEach(branchInfo => {
+    branchInfo.value.forEach((branchName, index) => 
+      !recycleTree.contains(branchName) &&
```
Comment (AndyCErnst) [inline infoTree/createRecycleTree.js]: this will be easier to read if you pull it out to a filter
```
branchInfo.value.filter( branchName => recycleTree.contains(branchName)).forEach((branchName, index) =>  
```

File: infoTree/createRecycleTree.js [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+  var recycleTree = new Tree()
+  recycleTree.add("rootNode")
+
+  //format object so its easier to use if it's not formatted
+  (typeof recycleArray[0].value==="string") && (recycleArray = formatRecycleArray(recycleArray))
+
+  recycleArray.forEach(branchInfo => {
+    branchInfo.value.forEach((branchName, index) => 
+      !recycleTree.contains(branchName) && 
+        (branchInfo.value[index-1] ? recycleTree.add(branchName, branchInfo.value[index-1]) : recycleTree.add(branchName, "rootNode")
```
Comment (AndyCErnst) [inline infoTree/createRecycleTree.js]: I'd separate this out a bit, something like
```
var infoToAdd = branchInfo.value[index-1] || "rootNode";
recycleTree.add(branchName, infoToAdd);
```

File: infoTree/createRecycleTree.js [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+
+
+
+
+
+
+//formatting whereDoesItGo. Changes the value string => array. 
+//This could be done with regular expressions, but an array containing the parents seems easier to me.
+//TODO: adapt for different inputs. because of the slice(4,-1) this could break easily when we include the category titles
+function formatRecycleArray(whereRecycleArray){
```
Comment (AndyCErnst) [inline infoTree/createRecycleTree.js:177]: This uses the tree structure and is not part of it, so I would put it in a different file.

File: infoTree/createDataStore.js [hunk 18]
```diff
@@ -0,0 +1,32 @@
+//Psuedocode for creating a dataStore for recycling information via scraping seattle.gov/wheredoesitgo
+//Mapping out things to be done and general pattern for data
+import createRecycleTree from './createRecycleTree';
+import whereDoesItGo from './whereDoesItGo';
+import nodeList from './nodeList';
+
+function scrapeSeattle(url){
+	//TODO: Figure out best methods for getting HTML from a page, probably have to fake some headers
+	// $.(get) ?
```
Comment (AndyCErnst) [inline infoTree/createDataStore.js]: $.get is a jQuery thing, fetch should have all the options we need

**Review by AndyCErnst — CHANGES_REQUESTED** [review]
A few code cleanliness things. It's looking good.
You should create some folders for these, maybe a data folder or an output folder.

File: infoTree/createDataStore.js (hunk 11 above)
Comment (adamosimpson21) [reply infoTree/createDataStore.js]: sweet! Thanks for showing me this

File: infoTree/createRecycleTree.js (hunk 14 above)
Comment (adamosimpson21) [reply infoTree/createRecycleTree.js]: I was just trying to be fancy. Probably too fancy

Comment (AndyCErnst) [reply infoTree/createRecycleTree.js]: lol, read this https://hackernoon.com/why-senior-devs-write-dumb-code-and-how-to-spot-a-junior-from-a-mile-away-27fa263b101a

File: infoTree/createRecycleTree.js (hunk 16 above)
Comment (adamosimpson21) [reply infoTree/createRecycleTree.js]: that's a great refactor :) thanks

File: infoTree/createRecycleTree.js (hunk 15 above)
Comment (adamosimpson21) [reply infoTree/createRecycleTree.js]: I'm having trouble with filter because the way I'm doing it, I think it initially filters before adding any of the branch elements, so it can't filter out what doesn't exist yet in the tree. I'm going to move the parts around and see if I can come up with anything

Comment (AndyCErnst) [reply infoTree/createRecycleTree.js]: ah, I missed that this is a dedupe. What happens if you try to add a duplicate? Does it overwrite it? When would there be duplicates?

Comment (adamosimpson21) [reply infoTree/createRecycleTree.js]: If you add a duplicate, it makes another node in the tree. We could add an internal structure in the tree to protect against duplicate branches, but It's possible you run into an issue with similar terms (cans, bottles, box, etc. can all be made of different materials). There are many duplicates since you're adding the 2nd level branches each time the forEach runs over them. And you can't exclude them, since the leaf is the information you're looking for

Comment (AndyCErnst) [reply infoTree/createRecycleTree.js]: In the final decision tree, it would make sense for there to have nodes of  "plastic" with child "box" and "cardboard" with child "box".
I haven't run your code though, I hope to have some time to take a closer look

File: infoTree/scrape.html [hunk 19]
```diff
@@ -0,0 +1,13 @@
...
+    <meta charset="UTF-8">
+    <title>title</title>
+  </head>
+  <body>
+  <button id="theButton">Button</button>
+  <div id="root" style="font-size:48px;"></div>
+  	
+  	<!-- <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script> -->
+  	<script src="./seattleGovScrape.js"></script>
+  </body>
```
Comment (AndyCErnst) [inline infoTree/scrape.html]: why is this file needed? You should be able to run the script in the console.

File: infoTree/seattleGovScrape.js [hunk 20]
```diff
@@ -52,73 +57,37 @@ var sampleObject= {
...
-]
-}
\ No newline at end of file
+
+function scrapeHtml(){
+	var frame = {
+		key:"AirConditionersHeatPumps",
+		title:"h1",
+	} 
+	var recycleInfo = $('#region').scrape(frame);
```
Comment (AndyCErnst) [inline infoTree/seattleGovScrape.js:66]: this API is interesting, I'll have to try this out. The jQuery-like syntax is nice but the scrape method is less straightforward.

**Review by AndyCErnst — COMMENTED** [review]
Ideally, this scraper works without a webpage, it's just a command that a person or a build pipeline would run and save the output to a file. The output files should be .json as they are json. That does mean they can't have comments though.

**Review by AndyCErnst — APPROVED** [review]
